### PR TITLE
Remove decryption from downloaded and change API

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

**Review by Eabramov84 — APPROVED** [review]

File: database/database.go [hunk 1]
```diff
@@ -301,6 +360,87 @@ func (db *Database) Close() {
  * Private
  **********************************************************************************************************************/
 
+func (db *Database) getDataFromQuery(query string, queryParams []any, result ...interface{}) error {
```
Comment (al1img) [inline database/database.go:363]: 1. Use either any or interface{}
2. If you introduce generic function, existing API should be changed to use it as well ;)

File: database/database.go [hunk 2]
```diff
@@ -301,6 +360,87 @@ func (db *Database) Close() {
...
+			return errNotExist
+		}
+
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (db *Database) getDownloadFromQuery(
```
Comment (al1img) [inline database/database.go:381]: What is the purpose of this function? Can be inside GetDownloadInfos.

File: database/database.go [hunk 3]
```diff
@@ -301,6 +360,87 @@ func (db *Database) Close() {
...
+			return nil, aoserrors.Wrap(err)
+		}
+
+		downloadInfos = append(downloadInfos, downloadInfo)
+	}
+
+	return downloadInfos, nil
+}
+
+func (db *Database) executeQuery(query string, args ...interface{}) error {
```
Comment (al1img) [inline database/database.go:409]: Update existing API to use this function as well.

File: database/database.go [hunk 4]
```diff
@@ -301,6 +360,87 @@ func (db *Database) Close() {
...
+	}
+
+	return nil
+}
+
+func (db *Database) createDownloadTable() (err error) {
+	log.Info("Create service table")
+
+	_, err = db.sql.Exec(`CREATE TABLE IF NOT EXISTS download (path TEXT NOT NULL PRIMARY KEY,
+                                                               target_type TEXT NOT NULL,
```
Comment (al1img) [inline database/database.go:437]: Use camelCase for field names.

File: database/database.go [hunk 5]
```diff
@@ -292,6 +297,60 @@ func (db *Database) GetSoftwareUpdateState() (state json.RawMessage, err error)
...
+
+	return err
+}
+
+func (db *Database) SetDownloadInfo(downloadInfo downloader.DownloadInfo) (err error) {
+	var path string
+
+	if err = db.getDataFromQuery(
+		"SELECT * FROM download WHERE path = ?",
+		[]any{downloadInfo.Path}, &path); err != nil {
```
Comment (al1img) [inline database/database.go:330]: Combine nested if with current one.

File: database/database.go [hunk 6]
```diff
@@ -292,6 +297,60 @@ func (db *Database) GetSoftwareUpdateState() (state json.RawMessage, err error)
...
+
+func (db *Database) RemoveDownloadInfo(filePath string) (err error) {
+	if err = db.executeQuery("DELETE FROM download WHERE path = ?", filePath); errors.Is(err, errNotExist) {
+		return nil
+	}
+
+	return err
+}
+
+func (db *Database) SetDownloadInfo(downloadInfo downloader.DownloadInfo) (err error) {
```
Comment (al1img) [inline database/database.go:325]: Still think that Create/Read/Update/Remove is better approach then create item in update as more common pattern (https://en.wikipedia.org/wiki/Create,_read,_update_and_delete). But up to you.

File: database/database.go [hunk 7]
```diff
@@ -292,6 +297,60 @@ func (db *Database) GetSoftwareUpdateState() (state json.RawMessage, err error)
...
+
+	if err = db.getDataFromQuery(
+		"SELECT * FROM download WHERE path = ?",
+		[]any{downloadInfo.Path}, &path); err != nil {
+		if !errors.Is(err, errNotExist) {
+			return err
+		}
+	}
+
+	if path != "" {
```
Comment (al1img) [inline database/database.go:336]: Checking errNotExist is more clear check then `path != ""`.

File: database/database.go [hunk 8]
```diff
@@ -301,6 +360,87 @@ func (db *Database) Close() {
...
+		return aoserrors.Wrap(err)
+	}
+
+	count, err := result.RowsAffected()
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if count == 0 {
+		return aoserrors.Wrap(errNotExist)
```
Comment (al1img) [inline database/database.go:427]: I guess internal error should not be wrapped.

File: database/database_internal_test.go [hunk 9]
```diff
@@ -228,3 +230,72 @@ func TestMultiThread(t *testing.T) {
...
+			t.Errorf("Can't set download info: %v", err)
+		}
+
+		downloadInfo, err := db.GetDownloadInfo(tCase.downloadInfo.Path)
+		if err != nil {
+			t.Errorf("Can't get download info: %v", err)
+		}
+
+		if !reflect.DeepEqual(downloadInfo, tCase.downloadInfo) {
+			t.Errorf("Unexpected download info")
```
Comment (al1img) [inline database/database_internal_test.go:268]: `Error`

File: database/database_internal_test.go [hunk 10]
```diff
@@ -228,3 +230,72 @@ func TestMultiThread(t *testing.T) {
...
+		}
+
+		downloadInfos, err := db.GetDownloadInfos()
+		if err != nil {
+			t.Errorf("Can't get list download info: %v", err)
```
Comment (al1img) [inline database/database_internal_test.go:273]: `download info list` or `list of download info`

File: downloader/downloader.go [hunk 11]
```diff
@@ -377,11 +456,33 @@ func (downloader *Downloader) getRequiredSize(fileName string, totalSize uint64)
...
+		}
+	}
+
+	if downloadInfo.Released {
+		return aoserrors.New("download file set as released")
+	}
+
+	result.interruptReason = downloadInfo.InterruptReason
+
+	if err = downloader.downloadPackage(result); err != nil && result.interruptReason == "" {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:472]: Done

File: downloader/downloader.go [hunk 12]
```diff
@@ -502,15 +603,14 @@ func (downloader *Downloader) download(url string, result *downloadResult) (err
...
+		log.WithFields(log.Fields{"url": url, "id": result.id, "reason": result.interruptReason}).Debug("Download resumed")
 
-		downloader.sender.SendAlert(downloader.prepareDownloadAlert(resp, result, "Download resumed reason: "+reason))
-
-		result.removeInterruptReason()
+		downloader.sender.SendAlert(downloader.prepareDownloadAlert(
+			resp, result, "Download resumed reason: "+result.interruptReason))
 	}
 
+	result.interruptReason = ""
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:612]: Done

File: downloader/downloader.go [hunk 13]
```diff
@@ -526,7 +626,7 @@ func (downloader *Downloader) download(url string, result *downloadResult) (err
 					"downloaded": resp.BytesComplete(), "reason": err,
 				}).Warn("Download interrupted")
 
-				result.storeInterruptReason(err.Error())
+				result.interruptReason = err.Error()
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:629]: Done

File: downloader/downloader.go [hunk 14]
```diff
@@ -163,10 +179,84 @@ func (downloader *Downloader) Download(
...
+		return err
+	}
+
+	downloadInfo.Released = true
+
+	if err := downloader.storage.SetDownloadInfo(downloadInfo); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	downloader.Lock()
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:233]: Done

File: downloader/downloader.go [hunk 15]
```diff
@@ -287,32 +373,21 @@ func (downloader *Downloader) acceptSpace(result *downloadResult) (err error) {
...
+	downloadInfos, err := downloader.storage.GetDownloadInfos()
 	if err != nil {
 		return aoserrors.Wrap(err)
 	}
 
-	for _, entry := range entries {
-		if setOutdatedErr := downloader.setItemOutdated(
-			filepath.Join(downloader.config.DownloadDir, entry.Name())); setOutdatedErr != nil && err == nil {
-			err = setOutdatedErr
+	for _, downloadInfo := range downloadInfos {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:382]: Done

File: downloader/downloader.go [hunk 16]
```diff
@@ -377,11 +456,33 @@ func (downloader *Downloader) getRequiredSize(fileName string, totalSize uint64)
 func (downloader *Downloader) process(result *downloadResult) error {
 	log.WithFields(log.Fields{"id": result.id}).Debug("Process download")
 
-	if err := downloader.downloadPackage(result); err != nil {
+	downloadInfo, err := downloader.storage.GetDownloadInfo(result.downloadFileName)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:460]: not relevant

File: database/database.go (hunk 6 above)
Comment (MykolaSolyanko) [reply database/database.go:325]: decided to leave it as it is

File: database/database.go (hunk 5 above)
Comment (MykolaSolyanko) [reply database/database.go:330]: Done

File: database/database.go (hunk 7 above)
Comment (MykolaSolyanko) [reply database/database.go:336]: Done

File: database/database.go (hunk 1 above)
Comment (MykolaSolyanko) [reply database/database.go:363]: 2. Done

File: database/database.go (hunk 2 above)
Comment (MykolaSolyanko) [reply database/database.go:381]: Remove

File: database/database.go (hunk 3 above)
Comment (MykolaSolyanko) [reply database/database.go:409]: Done

File: database/database.go (hunk 8 above)
Comment (MykolaSolyanko) [reply database/database.go:427]: in this case linter shows an error

File: database/database.go (hunk 4 above)
Comment (MykolaSolyanko) [reply database/database.go:437]: Done

File: database/database_internal_test.go (hunk 10 above)
Comment (MykolaSolyanko) [reply database/database_internal_test.go:273]: Done

File: downloader/downloader.go [hunk 17]
```diff
@@ -69,6 +65,14 @@ type Downloader struct {
 	currentDownloads map[string]*downloadResult
 	waitQueue        *list.List
 	allocator        spaceallocator.Allocator
+	storage          Storage
+}
+
+type DownloadInfo struct {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:71]: Done

File: downloader/downloader.go [hunk 18]
```diff
@@ -83,21 +87,33 @@ type PackageInfo struct {
 	TargetVendorVersion string
 }
 
+type Storage interface {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:90]: Done

File: downloader/downloader.go [hunk 19]
```diff
@@ -163,10 +179,84 @@ func (downloader *Downloader) Download(
...
 }
 
+func (downloader *Downloader) Release(filePath string) error {
+	downloadInfo, err := downloader.storage.GetDownloadInfo(filePath)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if downloadInfo.Released {
+		log.Warningf("file %v already released", filePath)
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go]: not relevant

File: downloader/downloader.go [hunk 20]
```diff
@@ -163,10 +179,84 @@ func (downloader *Downloader) Download(
...
+	}
+
+	return nil
+}
+
 /***********************************************************************************************************************
  * Private
  **********************************************************************************************************************/
 
+func (downloader *Downloader) setReleasedDownloadInfo(downloadInfo DownloadInfo) error {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go]: Done

File: downloader/downloader.go [hunk 21]
```diff
@@ -287,32 +355,40 @@ func (downloader *Downloader) acceptSpace(result *downloadResult) (err error) {
...
 		return aoserrors.Wrap(err)
 	}
 
+nextEntry:
 	for _, entry := range entries {
-		if setOutdatedErr := downloader.setItemOutdated(
-			filepath.Join(downloader.config.DownloadDir, entry.Name())); setOutdatedErr != nil && err == nil {
-			err = setOutdatedErr
+		downloadFilePath := filepath.Join(downloader.config.DownloadDir, entry.Name())
+		for _, downloadInfo := range downloadInfos {
```
Comment (al1img) [inline downloader/downloader.go:380]: Add empty line before for.

File: downloader/downloader.go [hunk 22]
```diff
@@ -287,32 +355,40 @@ func (downloader *Downloader) acceptSpace(result *downloadResult) (err error) {
...
+func (downloader *Downloader) setDownloadDirOutdated() error {
+	downloadInfos, err := downloader.storage.GetDownloadInfos()
+	if err != nil {
+		return aoserrors.Wrap(err)
 	}
 
-	if setOutdatedErr := downloader.setItemOutdated(result.interruptFileName); setOutdatedErr != nil && err == nil {
-		err = setOutdatedErr
+	for _, downloadInfo := range downloadInfos {
+		if !downloadInfo.Downloaded {
```
Comment (al1img) [inline downloader/downloader.go:365]: It could be inside next loop.

File: downloader/downloader.go [hunk 23]
```diff
@@ -287,32 +355,40 @@ func (downloader *Downloader) acceptSpace(result *downloadResult) (err error) {
...
-			filepath.Join(downloader.config.DownloadDir, entry.Name())); setOutdatedErr != nil && err == nil {
-			err = setOutdatedErr
+		downloadFilePath := filepath.Join(downloader.config.DownloadDir, entry.Name())
+		for _, downloadInfo := range downloadInfos {
+			if downloadFilePath == downloadInfo.Path {
+				continue nextEntry
+			}
+		}
+
+		if err := os.RemoveAll(downloadFilePath); err != nil {
```
Comment (al1img) [inline downloader/downloader.go:386]: You should not remove items here. It should remain cached. Only allocator should remove items when there is no space. This loop should look like:
```go
	for _, entry := range entries {
		downloadFilePath := filepath.Join(downloader.config.DownloadDir, entry.Name())

		for _, downloadInfo := range downloadInfos {
			if downloadFilePath == downloadInfo.Path && donwloadInfo.Downloaded {
				continue nextEntry
			}
		}

                if err := downloader.setItemOutdated(downloadFilePath); err != nil {
                       ...
                }
	}
```

File: downloader/downloader.go [hunk 24]
```diff
@@ -163,10 +183,60 @@ func (downloader *Downloader) Download(
...
+}
+
+func (downloader *Downloader) ReleaseByType(targetType string) error {
+	downloadInfos, err := downloader.storage.GetDownloadInfos()
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	for _, downloadInfo := range downloadInfos {
+		if downloadInfo.TargetType == targetType {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply downloader/downloader.go:210]: Done

File: downloader/downloader.go (hunk 21 above)
Comment (MykolaSolyanko) [reply downloader/downloader.go:380]: Done

File: downloader/downloader.go (hunk 22 above)
Comment (MykolaSolyanko) [reply downloader/downloader.go:365]: Done

File: downloader/downloader.go (hunk 23 above)
Comment (MykolaSolyanko) [reply downloader/downloader.go:386]: Done

File: downloader/downloader.go [hunk 25]
```diff
@@ -287,32 +357,39 @@ func (downloader *Downloader) acceptSpace(result *downloadResult) (err error) {
...
 
+nextEntry:
 	for _, entry := range entries {
-		if setOutdatedErr := downloader.setItemOutdated(
-			filepath.Join(downloader.config.DownloadDir, entry.Name())); setOutdatedErr != nil && err == nil {
-			err = setOutdatedErr
+		downloadFilePath := filepath.Join(downloader.config.DownloadDir, entry.Name())
+
+		for _, downloadInfo := range downloadInfos {
+			if !downloadInfo.Downloaded {
```
Comment (al1img) [inline downloader/downloader.go:376]: Why do you need this check on each iteration? What's wrong with proposed code:
```go
	for _, entry := range entries {
		downloadFilePath := filepath.Join(downloader.config.DownloadDir, entry.Name())

		for _, downloadInfo := range downloadInfos {
			if downloadFilePath == downloadInfo.Path && donwloadInfo.Downloaded {
				continue nextEntry
			}
		}

                if err := downloader.setItemOutdated(downloadFilePath); err != nil {
                       ...
                }
	}
```

**Review by al1img — APPROVED** [review]

File: downloader/downloader.go (hunk 25 above)
Comment (MykolaSolyanko) [reply downloader/downloader.go]: Done

**Pull request merged**